crates/btcio/src/reader/query.rs

```diff
@@ -258,16 +258,36 @@ async fn poll_for_new_blocks<R: Reader>(
     for fetch_height in scan_start_height..=client_height {
         match fetch_and_process_block(ctx, fetch_height, state, status_updates).await {
             Ok((blkid, evs)) => {
-                events.extend_from_slice(&evs);
+                // First handle events, i.e. send to csm
+                for ev in evs.iter() {
+                    handle_bitcoin_event(ev.clone(), ctx, event_submitter).await?;
+                }
 
+                // If there is checkpoint in blocks, then wait until that gets accepted or
+                // discarded i.e. if it is verified. If it is accepted, update filter config.
                 if let Some(checkpt) = find_checkpoint_in_events(&evs) {
                     // if we have a checkpoint in this block, update filterconfig based on this
-                    let chainstate = borsh::from_slice(checkpt.checkpoint().sidecar().chainstate())
-                        .expect("deserialize chainstate");
 
-                    state
-                        .filter_config_mut()
-                        .update_from_chainstate(&chainstate);
+                    let batch_info = checkpt.checkpoint().batch_info();
+                    let l2_commt = checkpt.checkpoint().batch_info().l2_range.1;
+                    let ckpt_epoch = batch_info.epoch();
+
+                    let span = info_span!("wait_checkpoint_accepted", %ckpt_epoch, last_blkid=%l2_commt.blkid());
+                    let accepted = wait_until_checkpoint_accepted_or_rejected(ctx, checkpt)
+                        .instrument(span)
+                        .await?;
+
+                    if accepted {
+                        if let Some(chainstate) = checkpt.checkpoint().sidecar_as_chainstate() {
+                            state
+                                .filter_config_mut()
+                                .update_from_chainstate(&chainstate);
+                        } else {
+                            anyhow::bail!("Could not extract chainstate from checkpoint");
```

Comment:
If `sidecar_as_chainstate` returned an error result then we wouldn't have to re-recognize the `None` condition and construct a new error.  We'd probably want to wrap it with additional context, but we could do that with `.map_err`.